[PATCH] Adding changes to write ip nht resolve-via-default

From: sudhanshukumar22 <sudhanshu.kumar@broadcom.com>
Date: Wed, 13 Nov 2024 23:12:22 -0800
Subject: [PATCH] Adding changes to write ip nht resolve-via-default
 configuration for VRF in zebra

---
 zebra/zebra_vrf.c |   34 ++++++++++++++++++++++++++++++++++
 zebra/zebra_vrf.h |    1 +
 zebra/zebra_vty.c |    2 ++
 3 files changed, 37 insertions(+)

diff --git a/zebra/zebra_vrf.c b/zebra/zebra_vrf.c
index 3231b03b8..3beb552c6 100644
--- a/zebra/zebra_vrf.c
+++ b/zebra/zebra_vrf.c
@@ -666,6 +666,40 @@ int zebra_vrf_netns_handler_create(struct vty *vty, struct vrf *vrf,
 	return CMD_SUCCESS;
 }
 
+void zebra_vrf_config_write(struct vty *vty)
+{
+	struct vrf *vrf;
+	struct zebra_vrf *zvrf;
+
+	RB_FOREACH (vrf, vrf_name_head, &vrfs_by_name) {
+		zvrf = vrf->info;
+
+		if (!zvrf)
+			continue;
+
+		if (zvrf_id(zvrf) == VRF_DEFAULT) {
+			if (zvrf->zebra_rnh_ip_default_route)
+				vty_out(vty, "ip nht resolve-via-default\n");
+
+			if (zvrf->zebra_rnh_ipv6_default_route)
+				vty_out(vty, "ipv6 nht resolve-via-default\n");
+		} else {
+			vty_frame(vty, "vrf %s\n", zvrf_name(zvrf));
+			if (zvrf->zebra_rnh_ip_default_route)
+				vty_out(vty, " ip nht resolve-via-default\n");
+
+			if (zvrf->zebra_rnh_ipv6_default_route)
+				vty_out(vty, " ipv6 nht resolve-via-default\n");
+		}
+		if (zvrf_id(zvrf) != VRF_DEFAULT)
+			vty_endframe(vty, "exit-vrf\n!\n");
+		else
+			vty_out(vty, "!\n");
+	}
+	return 0;
+
+}
+
 /* Zebra VRF initialization. */
 void zebra_vrf_init(void)
 {
diff --git a/zebra/zebra_vrf.h b/zebra/zebra_vrf.h
index f37e7dcc3..3bd100970 100644
--- a/zebra/zebra_vrf.h
+++ b/zebra/zebra_vrf.h
@@ -252,6 +252,7 @@ extern vrf_id_t zebra_vrf_lookup_by_table(uint32_t table_id, ns_id_t ns_id);
 extern struct zebra_vrf *zebra_vrf_alloc(struct vrf *vrf);
 extern struct route_table *zebra_vrf_table(afi_t, safi_t, vrf_id_t);
 int zebra_vrf_lookup_tableid(vrf_id_t vrf_id, ns_id_t ns_id);
+extern void zebra_vrf_config_write(struct vty *vty);
 
 /*
  * API to associate a VRF with a NETNS.
diff --git a/zebra/zebra_vty.c b/zebra/zebra_vty.c
index b4e35e49b..3fa935484 100644
--- a/zebra/zebra_vty.c
+++ b/zebra/zebra_vty.c
@@ -3702,6 +3702,8 @@ static int config_write_protocol(struct vty *vty)
 
 	zebra_pbr_config_write(vty);
 
+	zebra_vrf_config_write(vty);
+
 	if (!zebra_vxlan_get_accept_bgp_seq())
 		vty_out(vty, "no evpn accept-bgp-seq\n");
 
